Docs: Guiding Principles (#51)
advice-animal/ick: docs/concepts/guiding_principles.md

> # Guiding Principles for ick

Reviewer: ```suggestion
# Guiding Principles for Ick
```

> Pithiily,

Reviewer: ```suggestion
Pithily,
```
Three alternate solutions are
1. Ultimately
2. In short
3. TL;DR

* First, corrupt no source code.
* Second, don't reinvent what works.

## Explicit Goals

* Always give the user an "undo" (default to an informative dry run)
* Be something that one person on a team can adopt, if they're the only one that cares
* Be something that you can run infrequently (say, before a release or in a tech

>   debt week), while resepcting the amount of time you have available

Reviewer: ```suggestion
  debt week), while respecting the amount of time you have available
```

> * Don't be tied to any one {language,os,etc}: Be agnostic to what tools people

Reviewer: ```suggestion
* Don't be tied to any one language, operating system, or similar constraint: Be agnostic to what tools people
```

  want to use
* Give you a heads-up about recommendations that you don't need to apply yet
  (but can if you have spare time or want to be an early adopter)

> By accomodating a wide range of levels of effort, we can apply automation where

Reviewer: ```suggestion
By accommodating a wide range of levels of effort, we can apply automation where
```

Reviewer: Same here, almost didn't notice. Kudos to Tim for the idea of reading everything backwards.

> # Guiding Principles

Reviewer 2: ```suggestion
# Guiding Principles for ick
```

First, the pithy version:

* First, corrupt no source code.
* Second, don't reinvent what works.

## Explicit Goals

> * Always give you an "undo" (default to an informative dry run)

Reviewer 2: ```suggestion
* Always give the user an "undo" (default to an informative dry run)
```

* Be something that one person on a team can adopt, if they're the only one that cares
* Be something that you can run infrequently (say, before a release or in a tech
  debt week), while resepcting the amount of time you have available
* Don't be tied to any one {language,os,etc}: Be agnostic to what tools people
  want to use
* Give you a heads-up about recommendations that you don't need to apply yet
  (but can if you have spare time or want to be an early adopter)

> By focusing on a sliding scale of effort, we can apply automation where it

Reviewer 2: ```suggestion
By accommodating a wide range of levels of effort, we can apply automation where it
```

> By focusing on a sliding scale of effort, we can apply automation where it

Reviewer: I liked Ned's suggestion of `By accommodating a variety of scales of effort...`

makes sense, especially on the low end of complexity.  If you haven't wanted to

> write a custom flake8 plugin because it's hard, and resorted to a `grep`

Reviewer 2: ```suggestion
write a custom linter plugin because it's hard, and resorted to a `grep`
```

Reviewer 2: Trying to steer away from things only Python people will know.

somewhere, this framework is for you.

## On-Ramps

To make it easier to try or, or for only certain people to use at the
beginning, it must be possible to use `ick` in an existing repo with the most
minimal configuration possible.

You should also be able to script `ick`, running it on directories and doing
your own operations on the output.  `Ick` won't be creating you PRs using the
GitHub API, but you should be able to build such an automation for yourself or
your company.

You should be able to run `ick` with point-in-time rules from something like a
Sourcegraph Batch Change, which means it should be easy to package into a
Docker image with your own custom rules.

## Off-Ramps

You should be able to stop using `ick` with only things users have explicitly

> done (like suppressions or ignores) left behind.

Reviewer 2: I've heard this a few times, but don't quite understand.  In what way do other tools make it hard to stop using the tool?

Author: There's no subscription, no dependabot.yml, no WORKSPACE, no pre-pyproject.toml configs cluttering up the root.
